### PR TITLE
Integrate FED25 handling in electron online reconstruction

### PR DESCRIPTION
**Pull request by RSalvatico**
#### PR description:

This PR contains changes needed for the electron online reconstruction to be able to handle errors such as FED25 to flag malfunctioning Pixel modules. With respect to the current way of handling errors, which is maintained via the `isActive()` method

https://github.com/cms-sw/cmssw/blob/754d4fbc8bf3e05df6de16937fe6f4d30f5d8699/RecoEgamma/EgammaElectronAlgos/src/TrajSeedMatcher.cc#L356

the new method allows us to catch cases where the flagging happens on a per-event basis (i.e., no need to mask those modules permanently via SiPixelQualityRcd).

In addition, if in the future a 2023 MC will be made where the FED25 error is used to simulate the BPIX issue (e.g., via a probability function), with these changes the online electron reconstruction code should be able to handle it correctly.

#### PR validation:

I observed an improvement in efficiency for electron triggers (such as HLT_Ele32_WPTight_Gsf) when running them with the proposed changes. The sample used for the test was a DY simulation where the issue in the faulty BPIX modules in layers 3 and 4 in 2023 was simulated at the level of the digitizer (i.e., the channels corresponding to the faulty modules produce an error FED25). Some plots can be found [here](https://cernbox.cern.ch/s/iB7yjoSbo3ne3Uf). Please note that the efficiencies are calculated using gen-level electrons and basically no selection, so on an absolute scale they can look low.

The PR also passed some basic runTeMatrix tests.

### CODE REVIEW

**Comment by RSalvatico** [comment]
type egamma

**Comment by RSalvatico** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by mmusich** [comment]
> This PR contains changes needed for the electron online reconstruction to be able to handle errors such as FED25 to flag malfunctioning Pixel modules.

out of curiosity, has this any measurable impact on the HLT timing? From the [PR report](https://github.com/cms-sw/cmssw/pull/43555#issuecomment-1852796051) looks like we're running the on demand Strip at HLT clustering more often with these changes (I guess because of more electron seeds?)

**Comment by RSalvatico** [comment]
> > This PR contains changes needed for the electron online reconstruction to be able to handle errors such as FED25 to flag malfunctioning Pixel modules.
> 
> out of curiosity, has this any measurable impact on the HLT timing? From the [PR report](https://github.com/cms-sw/cmssw/pull/43555#issuecomment-1852796051) looks like we're running the on demand Strip at HLT clustering more often with these changes (I guess because of more electron seeds?)

Yes, it makes sense that there is some increase. I can measure the timing.

File: RecoTracker/MeasurementDet/plugins/TkPixelMeasurementDet.cc [hunk 1]
```diff
@@ -153,19 +153,24 @@ bool TkPixelMeasurementDet::hasBadComponents(const TrajectoryStateOnSurface& tso
     return false;
 
   auto lp = tsos.localPosition();
-  auto le = tsos.localError().positionError();
+  float xx = 0.;
+  float yy = 0.;
+  if (tsos.hasError()) {
+    xx = tsos.localError().positionError().xx();
+    yy = tsos.localError().positionError().yy();
+  }
```
Comment (mmusich) [inline RecoTracker/MeasurementDet/plugins/TkPixelMeasurementDet.cc:161]:
```suggestion
  float xx = tsos.hasError() ? tsos.localError().positionError().xx() : 0.0f;
  float yy = tsos.hasError() ? tsos.localError().positionError().yy() : 0.0f;
```

is slightly more compact.

**Review by mmusich — COMMENTED** [review]
some unrequested (and un-necessary) feedback below.
Also somehow `le` is a widespread shorthand in tracking code for local error, if the new variables were called `lexx` and `leyy` perhaps the code would be slightly less obfuscated.

**Comment by RSalvatico** [comment]
@mmusich The impact of these changes on the HLT timing shouldn't be large (see attached plot). I've run it another time and saw a slightly larger difference (~2 ms increase rather than the 0.4 ms you see here), so I guess it's within uncertainties.

Screenshot 2023-12-14 alle 14 37 37

**Comment by mmusich** [comment]
> The impact of these changes on the HLT timing shouldn't be large (see attached plot). I've run it another time and saw a slightly larger difference (~2 ms increase rather than the 0.4 ms you see here), so I guess it's within uncertainties.

thank you @RSalvatico

**Comment by jfernan2** [comment]
+1

**Comment by rappoccio** [comment]
+1

**Pull request merged**